### PR TITLE
Prevent CLI to panic when the pipe is broken, fixes #5

### LINKED ISSUE #5: Prevent CLI to panic when the pipe is broken
https://github.com/rust-lang/rust/issues/46016#issuecomment-1242039016

` carbonintensity-api -s 2023-11-11 postcode bs7 | head
`

### PR DESCRIPTION
**Pull request by jnioche**

### CODE REVIEW

**Comment by xoen** [comment]
I'm trying to wrap my head around this...not quite the diff itself which is _**relatively**_ "clear" (use standard C library's [`signal()` function](https://www.man7.org/linux/man-pages/man2/signal.2.html) to "set the disposition of the signal" to use the default action)

But I'm trying to figure out why this happens, this [SO answer](https://stackoverflow.com/a/44865822/455642) is the best explanation I've found so far.

TL;DR: it's to do with how `println!()` expands to use `::std::io::_print()` which can panic (the answers point to the reasons why but it doesn't really matter)

...but it's complicated and it shouldn't be! We write Rust so that we can avoid the horrible pitfalls of C most of the times...using the standard C library directly defeats the purpose of this :)

**The problem really is that `println!()` fails to get hold of stdout and it fails.**

But there is an alternative, use `writeln!()`, which returns a `Result<T>` instead of panicking!
This makes the "ignore write error" behaviour more explicit, for example if you print using `writeln!()` like this:

```Rust
// note that `writeln!()` uses the `std::io::Write` trait which has to be in scope
writeln!(std::io::stdout(), "{}...", whatever);
```

Rust would warn you that `writeln!()` may fail, it returns a `Result` and the potential error should be handled explicitly.
If you just `unwrap()` you get roughly the same behaviour as before (panic) but in this case you could use `unwrap_or_default()` to ignore the error and carry on.

In other words replace the `println!` in `main.rs` with `writeln!(std::io::stderr(), [...]` should solve the issue (unless I'm misunderstanding this, which is possible of course) without having to rely on conditional compilation, foreign functions, unstable, the standard C library being present and what not :) (which has the added benefit of possibly solving this problem on any non-unix platform)

Don't know if any of this makes sense :)

**Comment by jnioche** [comment]
Clippy doesn't like it 
https://rust-lang.github.io/rust-clippy/master/index.html#explicit_write

We'll tell it to shut up

File: src/main.rs [hunk 1]
```diff
@@ -41,20 +42,28 @@ async fn main() {
...
 fn handle_result(result: Result<i32, ApiError>, target: &Target) {
     if result.is_ok() {
-        println!("Carbon intensity for {}: {:?}", target, result.unwrap());
+        writeln!(
+            std::io::stdout(),
+            "Carbon intensity for {}: {:?}",
+            target,
+            result
+        )
+        .unwrap();
```
Comment (xoen) [inline src/main.rs:66]: this should be `unwrap_or_default()` also, otherwise it would still panic :)

**Comment by xoen** [comment]
Possibly worthy adding a note to the CHANGELOG for this?

```Markdown
### Fixed

- [...]
- prevent CLI to panic when the pipe is broken ([issue #5](https://github.com/jnioche/carbonintensity-api/issues/5))
```

**Review by xoen — APPROVED** [review]
🎉

**Pull request merged**